pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -210,19 +231,134 @@ pub(super) fn spawn_storage_sync_thread<
+fn storage_sync_loop<
+    P: std::fmt::Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+>(
     config: &'static PageServerConf,
     remote_storage: S,
-    concurrent_sync_limit: &Semaphore,
+    max_concurrent_sync: NonZeroUsize,
+    max_sync_errors: NonZeroU32,
 ) -> anyhow::Result<()> {
     let runtime = tokio::runtime::Builder::new_current_thread()
         .enable_all()
         .build()?;
-    let mut remote_timelines = runtime
+    let remote_timelines = runtime
         .block_on(fetch_existing_uploads(&remote_storage))
         .context("Failed to determine previously uploaded timelines")?;
 
-    let urgent_downloads = latest_timelines(&remote_timelines)
+    schedule_first_tasks(config, &remote_timelines);
+
+    let remote_timelines = Arc::new(Mutex::new(remote_timelines));
+    let remote_storage = Arc::new(remote_storage);
+
+    while !crate::tenant_mgr::shutdown_requested() {
+        let next_tasks = get_next_tasks(max_concurrent_sync);
+        let remaining_queue_length = SYNC_QUEUE.len();
+
+        if next_tasks.is_empty() {
+            trace!("No storage sync tasks found");
+            thread::sleep(Duration::from_secs(1));
+            continue;
+        } else {
+            debug!(
+                "Processing {} tasks, more tasks left to process: {}",
+                next_tasks.len(),
+                remaining_queue_length
+            );
+            REMAINING_SYNC_ITEMS.set(remaining_queue_length as i64);
+
+            let mut task_batch = next_tasks
+                .into_iter()
+                .map(|task| {
+                    tokio::spawn(process_task(
+                        config,
+                        Arc::clone(&remote_storage),
+                        task,
+                        max_sync_errors,
+                        Arc::clone(&remote_timelines),
+                    ))
+                })
+                .collect::<FuturesUnordered<_>>();
+            runtime.block_on(async move {
+                while let Some(task) = task_batch.next().await {
+                    if let Err(e) = task {
+                        error!("Failed to schedule storage sync task: {:#}", e);
+                    }
+                }
+            });
+        }
+    }
+    debug!("Queue based remote storage sync thread shut down");
+    Ok(())
+}
+
+async fn process_task<
+    P: std::fmt::Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+>(
+    config: &'static PageServerConf,
+    remote_storage: Arc<S>,
+    task: SyncTask,
+    max_sync_errors: NonZeroU32,
+    remote_timelines: Arc<Mutex<HashMap<(ZTenantId, ZTimelineId), RemoteTimeline>>>,
+) {
+    if task.retries > max_sync_errors.get() {
+        error!(
+            "Evicting task {:?} that failed {} times, exceeding the error theshold",
+            task.kind, task.retries
+        );
```

Comment:
Does unloading the entire timeline make sense if a task for it fails?

Reply:
Yes, same question here, not sure: it's lame to unload the data from the network outage indeed.

But what to do with tasks that fail for real then? We cannot allow such to accumulate, but presumably we cannot just skip one particular task and allow syncing the same timeline further later: if it goes for the upload task, missing a bunch of checkpoint files and allowing uploading another checkpoint later could render the whole timeline corrupt if I get it right.

So, we have to prohibit sync at least. That's doable, every `LayeredTimeline` carries the `uupload_relishes` flag inside that can be changed. But how to indicate that this timeline is not ok? This could be a big bad surprise too, while unloaded timeline is at least explicit.

Reply:
So far, let's keep it this way, we should load test this all eventually, could test some failure scenarios along the way too.